test/e2e/reconciler/workspace/controller_test.go

```
@@ -43,78 +44,103 @@ func TestWorkspaceController(t *testing.T) {
-		expect      framework.RegisterWorkspaceExpectation
-		expectShard framework.RegisterWorkspaceShardExpectation
+		rootKcpClient, orgKcpClient   clientset.Interface
+		rootKubeClient, orgKubeClient kubernetesclientset.Interface
+		orgExpect                     framework.RegisterClusterWorkspaceExpectation
+		rootExpectShard               framework.RegisterWorkspaceShardExpectation
 	}
 	var testCases = []struct {
 		name string
 		work func(ctx context.Context, t *testing.T, server runningServer)
 	}{
 		{
-			name: "add a shard after a workspace is unschedulable, expect it to be scheduled",
+			name: "create a workspace without shards, expect it to be unschedulable",
 			work: func(ctx context.Context, t *testing.T, server runningServer) {
-				workspace, err := server.client.TenancyV1alpha1().Workspaces().Create(ctx, &tenancyv1alpha1.Workspace{ObjectMeta: metav1.ObjectMeta{Name: "steve"}, Status: tenancyv1alpha1.WorkspaceStatus{}}, metav1.CreateOptions{})
+				t.Logf("Delete all pre-configured shards, we have to control the creation of the workspace shards in this test")
+				err := server.rootKcpClient.TenancyV1alpha1().WorkspaceShards().DeleteCollection(ctx, metav1.DeleteOptions{}, metav1.ListOptions{})
+				require.NoError(t, err)
+
+				t.Logf("Create a workspace without shards")
+				workspace, err := server.orgKcpClient.TenancyV1alpha1().ClusterWorkspaces().Create(ctx, &tenancyv1alpha1.ClusterWorkspace{ObjectMeta: metav1.ObjectMeta{Name: "steve"}, Status: tenancyv1alpha1.ClusterWorkspaceStatus{}}, metav1.CreateOptions{})
 				require.NoError(t, err, "failed to create workspace")
+				server.Artifact(t, func() (runtime.Object, error) {
+					return server.orgKcpClient.TenancyV1alpha1().ClusterWorkspaces().Get(ctx, workspace.Name, metav1.GetOptions{})
+				})
+
+				t.Logf("Expect workspace to be unschedulable")
+				if err := server.orgExpect(workspace, unschedulable); err != nil {
+					t.Errorf("did not see workspace marked unschedulable: %v", err)
```

This won't ensure the test fails fast. Maybe switch to t.Fatal or require.* ?